From: Dermot Bradley <dermot_bradley@yahoo.com>
Date: Sun,  8 Jan 2023 19:13 +0000
Subject: [PATCH] Fix various issues

Backport of changes made in ca-certificates-20230106 upstream source
package:

Modify c_rehash to ignore the /etc/ssl/certs/certs/ca-certificates.crt
file when created hash softlinks in the /etc/ssl/certs/ directory.
There are 3 reasons for this: (1) normally whenever
"update-ca-certificates" is run (whether by a script/trigger or by an
end-user) a warning will appear:

  WARNING: ca-certificates.crt does not contain exactly one certificate
  or CRL: skipping

which is annoying but not problematic, however (2) in some circumstances
such as where that file only contains a single certificate, i.e. where
you disable *all* the system certificates and only use your own "local"
CA file, then c_rehash may wrongly create a softlink to the
ca-certificates.crt file rather than to the correct CA file if c_rehash
finds the ca-certificates.crt file first in the /etc/ssl/certs/
directory.

In the "single CA cert" situation if however c_rehash finds the correct
(single CA) file first then (3) when it does find the ca-certificates.crt
file it will generate the following warning:

  WARNING: Skipping duplicate certificate in file ca-certificates.crt

The changes in this MR prevent all 3 scenarios from occurring.

Remove the blacklist.txt file - this has not been updated since it was
added to the repo 5+ years ago. Also this file is not used for any
purpose - it was previously used/read by the certdata2pem.py Python
script but that was replaced by by the mk-ca-bundle.pl Perl script
which makes no use of blacklist.txt.

Correct update-ca-certificates manpage - this was copied from Debian,
however the Alpine program does NOT support any cli options, so remove
these from the manpage.

---

diff -Naur a/blacklist.txt b/blacklist.txt
--- a/blacklist.txt
+++ b/blacklist.txt
@@ -1,23 +0,0 @@
-# One blacklist entry per line, corresponding to the label in certdata.txt.
-
-# MD5 Collision Proof of Concept CA
-"MD5 Collisions Forged Rogue CA 25c3"
-
-# DigiNotar Root CA (see debbug#639744)
-"DigiNotar Root CA"
-
-# StartCom and WoSign certificates are now untrusted by the major browser
-# vendors[0]. See [1] for discussion. The list was generated by:
-#
-#   $ egrep 'WoSign|StartCom' mozilla/certdata.txt \
-#         | grep UTF | sed 's/CKA_LABEL UTF8 //' | uniq
-#
-# [0] https://blog.mozilla.org/security/2016/10/24/distrusting-new-wosign-and-startcom-certificates/
-# [1] https://bugs.debian.org/858539
-#
-"StartCom Certification Authority"
-"StartCom Certification Authority G2"
-"WoSign"
-"WoSign China"
-"Certification Authority of WoSign G2"
-"CA WoSign ECC Root"
diff -Naur a/c_rehash.c b/c_rehash.c
--- a/c_rehash.c
+++ b/c_rehash.c
@@ -253,6 +253,11 @@
 			continue;
 		if (S_ISLNK(st.st_mode) && handle_symlink(de->d_name, buf) == 0)
 			continue;
+		if (strcmp(buf, "/etc/ssl/certs/ca-certificates.crt") == 0) {
+			/* Ignore the /etc/ssl/certs/ca-certificates.crt file */
+			if (do_verbose) printf("Skipping %s file\n", buf);
+			continue;
+		}
 		handle_certificate(de->d_name, buf);
 	}
 	closedir(d);
diff -Naur a/update-ca-certificates.8 b/update-ca-certificates.8
--- a/update-ca-certificates.8
+++ b/update-ca-certificates.8
@@ -19,7 +19,6 @@
 update-ca-certificates \- update /etc/ssl/certs and ca-certificates.crt
 .SH SYNOPSIS
 .B update-ca-certificates
-.RI [ options ]
 .SH DESCRIPTION
 This manual page documents briefly the
 .B update-ca-certificates
@@ -41,17 +40,6 @@
 .PP
 Before terminating, \fBupdate-ca-certificates\fP invokes
 \fBrun-parts\fP on /etc/ca-certificates/update.d.
-.SH OPTIONS
-A summary of options is included below.
-.TP
-.B \-h, \-\-help
-Show summary of options.
-.TP
-.B \-v, \-\-verbose
-Be verbose. Output \fBc_rehash\fP.
-.TP
-.B \-f, \-\-fresh
-Fresh updates.  Remove symlinks in /etc/ssl/certs directory.
 .SH FILES
 .TP
 .I /etc/ca-certificates.conf
@@ -69,5 +57,5 @@
 .SH SEE ALSO
 .BR c_rehash (1)
 .SH AUTHOR
-This manual page was written by Fumitoshi UKAI <ukai@debian.or.jp>,
-for the Debian project (but may be used by others).
+This manual page is based on the one written by Fumitoshi UKAI
+<ukai@debian.or.jp>, for the Debian project.
